### PR TITLE
DOS: Scala script and dockerfile

### LINKED ISSUE #3834: Scala scripts for DOS

### PR DESCRIPTION
**Pull request by salonishah11**
This PR adds Scala script and Dockerfile needed for resolving DOS Url and localizing it to the container. 

Closes #3834

### CODE REVIEW

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def resolveDosThroughMartha(dosUrl: String) : Try[MarthaResponse] = {
+  import MarthaResponseJsonSupport._
+
+  // if using fake Martha request, insert actual service account json here
+  // if using actual Martha remove this variable entirely
+  val serviceAccount = raw"""{}"""
+
+  val requestBody = json"""{"dosUrl":$dosUrl}"""
```
Comment (dvoet) [inline scripts/docker-dos/dosUrlLocalizer.sc]: this should be `url` not `dosUrl`

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import org.http4s._
+import org.http4s.circe._
+import io.circe.literal._
+import io.circe.syntax._
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+import scala.util.{Try, Success, Failure}
+
+
+case class MarthaResponse(dos: String, sa: String)
```
Comment (dvoet) [inline scripts/docker-dos/dosUrlLocalizer.sc]: this is not right but we discussed that already

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  // if using actual Martha remove this variable entirely
+  val serviceAccount = raw"""{}"""
+
+  val requestBody = json"""{"dosUrl":$dosUrl}"""
+
+  Try {
+    val httpClient = Http1Client[IO]().unsafeRunSync
+    //request to fake Martha
+    val postRequest = Request[IO](
+      method = Method.POST,
```
Comment (dvoet) [inline scripts/docker-dos/dosUrlLocalizer.sc]: need to add headers here with the authorization header

Comment (dvoet) [reply scripts/docker-dos/dosUrlLocalizer.sc]: maybe with something like
```
val credentials = GoogleCredentials getApplicationDefault()
credentials.refreshToken()
val accessToken = credentials.getAccessToken()
```
then the header is Authorization with a value of `s"bearer $accessToken"`

File: scripts/docker-dos/dosUrlLocalizer.sc (hunk 1 above)
Comment (salonishah11) [reply scripts/docker-dos/dosUrlLocalizer.sc]: I was using the function set up by Dan till now, which was expecting `dosUrl`. Will be changing it to `url` for the cloud function you set up. 👍

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  for {
+    credentials <- Try(GoogleCredentials.fromStream(new ByteArrayInputStream(serviceAccount.getBytes()))
+      .createScoped(Lists.newArrayList("https://www.googleapis.com/auth/cloud-platform")))
+    storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService()
+    blob <- Try(storage.get(gcsBucket, fileToBeLocalized))
+    readChannel <- Try(blob.reader())
+    _ <- Try(Files.createDirectories(Paths.get(downloadLoc).getParent))
+    fileOuputStream = new FileOutputStream(downloadLoc)
+    _ <- Try(fileOuputStream.getChannel().transferFrom(readChannel, 0, Long.MAX_VALUE))
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply scripts/docker-dos/dosUrlLocalizer.sc]: 👍

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  }
+}
+
+
+def resolveDosThroughMartha(dosUrl: String) : Try[MarthaResponse] = {
+  import MarthaResponseJsonSupport._
+
+  // if using fake Martha request, insert actual service account json here
+  // if using actual Martha remove this variable entirely
+  val serviceAccount = raw"""{}"""
```
Comment (mcovarr) [inline scripts/docker-dos/dosUrlLocalizer.sc]: remove entirely?

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  val requestBody = json"""{"dosUrl":$dosUrl}"""
+
+  Try {
+    val httpClient = Http1Client[IO]().unsafeRunSync
+    //request to fake Martha
+    val postRequest = Request[IO](
+      method = Method.POST,
+      uri = Uri.uri("https://us-central1-broad-dsde-cromwell-dev.cloudfunctions.net/helloWorld"))
+      .withBody(requestBody)
+    val httpResponse = httpClient.expect[String](postRequest).unsafeRunSync
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply scripts/docker-dos/dosUrlLocalizer.sc]: The code snippet worked like a charm! `IO[_]` is very very new to me as well. Thanks for pointing it out! :)

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    credentials <- Try(GoogleCredentials.fromStream(new ByteArrayInputStream(serviceAccount.getBytes()))
+      .createScoped(Lists.newArrayList("https://www.googleapis.com/auth/cloud-platform")))
+    storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService()
+    blob <- Try(storage.get(gcsBucket, fileToBeLocalized))
+    readChannel <- Try(blob.reader())
+    _ <- Try(Files.createDirectories(Paths.get(downloadLoc).getParent))
+    fileOuputStream = new FileOutputStream(downloadLoc)
+    _ <- Try(fileOuputStream.getChannel().transferFrom(readChannel, 0, Long.MAX_VALUE))
+    _ <- Try(fileOuputStream.close())
+  } yield()
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply scripts/docker-dos/dosUrlLocalizer.sc]: I was going back and forth between for comprehension and giant Try. Will change it to Try. Thanks!

File: scripts/docker-dos/dosUrlLocalizer.sc (hunk 5 above)
Comment (salonishah11) [reply scripts/docker-dos/dosUrlLocalizer.sc]: 👍

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  val dosResloverObj = for {
+    marthaResObj <- resolveDosThroughMartha(dosUrl)
+    _ <- downloadFileFromGcs(marthaResObj.dos, marthaResObj.sa, downloadLoc)
+  } yield()
+
+  dosResloverObj match {
+    case Success(_) =>
+    case Failure(e) => {
+      println("Error:")
+      println(e.printStackTrace)
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply scripts/docker-dos/dosUrlLocalizer.sc]: Thanks! 👍

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+case class DosObject(data_object: DosDataObject)
+
+case class MarthaResponse(dos: DosObject, googleServiceAccount: JsObject)
+
+
+@main
+def dosUrlResolver(dosUrl: String, downloadLoc: String) : Unit = {
+  val dosResloverObj = for {
+    marthaResObj <- resolveDosThroughMartha(dosUrl)
+    _ <- downloadFileFromGcs(marthaResObj.dos.data_object.urls(0).url, marthaResObj.googleServiceAccount.toString, downloadLoc)
```
Comment (dvoet) [inline scripts/docker-dos/dosUrlLocalizer.sc]: find the first instance of a url starting with `gs://` and throw an exception if none found

File: scripts/docker-dos/dosUrlLocalizer.sc [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  val credentials = GoogleCredentials.getApplicationDefault()
+  credentials.refreshAccessToken()
+  val accessToken = credentials.getAccessToken()
+
+  val marthaResponseIo: IO[MarthaResponse] = for {
+    httpClient <- Http1Client[IO]()
+    //request to fake Martha
+    postRequest <- Request[IO](
+      method = Method.POST,
+      uri = Uri.uri("http://us-central1-dvoet-mock-marthav2.cloudfunctions.net/martha_v2"),
```
Comment (dvoet) [inline scripts/docker-dos/dosUrlLocalizer.sc]: I assume this URL will come from configat some point

File: scripts/docker-dos/dosUrlLocalizer.sc (hunk 9 above)
Comment (salonishah11) [reply scripts/docker-dos/dosUrlLocalizer.sc]: 👍

**Pull request merged**